File: app/src/main/java/ch/bailu/aat/gpx/GpxPointNode.java
package ch.bailu.aat.gpx;

import androidx.annotation.NonNull;

import ch.bailu.aat.coordinates.BoundingBoxE6;
import ch.bailu.aat.gpx.attributes.GpxAttributes;
import ch.bailu.aat.gpx.attributes.Keys;
import ch.bailu.aat.gpx.interfaces.GpxDeltaPointInterface;
import ch.bailu.aat.gpx.linked_list.Node;


public abstract class GpxPointNode extends Node implements GpxDeltaPointInterface {

    private final GpxAttributes attributes;
    private final GpxPoint     point;


    public GpxPointNode(GpxPoint tp, GpxAttributes at) {
        point = tp;
        attributes = at;
    }


    public double getAltitude() {
        return point.getAltitude();
    }


    public double getLatitude() {
        return point.getLatitude();
    }


    public double getLongitude() {
        return point.getLongitude();
    }


    public long getTimeStamp() {
        return point.getTimeStamp();
    }


    public int getLatitudeE6() {
        return point.getLatitudeE6();
    }


    public int getLongitudeE6() {
        return point.getLongitudeE6();
    }


    @NonNull
    @Override
    public String toString() {
        return attributes.toString();
    }

    public GpxPoint getPoint() {
        return point;
    }

    public GpxAttributes getAttributes() {
        return attributes;
    }


    public void setAltitude(double e) {
        point.setAltitude((float) e);
    }


    private static final int BOUNDING_KEY= Keys.toIndex("boundingbox");


    @Override
    public BoundingBoxE6 getBoundingBox() {
        final BoundingBoxE6 box = new BoundingBoxE6();

        if (attributes.hasKey(BOUNDING_KEY)) {
            box.add(attributes.get(BOUNDING_KEY));
        }
        box.add(getPoint());

        return box;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/gpx/GpxPointNode.kt
package ch.bailu.aat_lib.gpx

import ch.bailu.aat_lib.coordinates.BoundingBoxE6
import ch.bailu.aat_lib.gpx.attributes.GpxAttributes
import ch.bailu.aat_lib.gpx.attributes.Keys.Companion.toIndex
import ch.bailu.aat_lib.gpx.interfaces.GpxDeltaPointInterface
import ch.bailu.aat_lib.gpx.linked_list.Node
import javax.annotation.Nonnull

abstract class GpxPointNode(@JvmField val point: GpxPoint, private val attributes: GpxAttributes) : Node(),
    GpxDeltaPointInterface {
    override fun getAltitude(): Float {
        return point.getAltitude()
    }

    override fun getLatitude(): Double {
        return point.getLatitude()
    }

    override fun getLongitude(): Double {
        return point.getLongitude()
    }

    override fun getTimeStamp(): Long {
        return point.getTimeStamp()
    }

    override fun getLatitudeE6(): Int {
        return point.getLatitudeE6()
    }

    override fun getLongitudeE6(): Int {
        return point.getLongitudeE6()
    }

    @Nonnull
    override fun toString(): String {
        return attributes.toString()
    }

    override fun getAttributes(): GpxAttributes {
        return attributes
    }

    fun setAltitude(e: Double) {
        point.setAltitude(e.toFloat())
    }

    override fun getBoundingBox(): BoundingBoxE6 {
        val box = BoundingBoxE6()

        if (attributes.hasKey(BOUNDING_KEY)) {
            box.add(attributes[BOUNDING_KEY])
        }
        box.add(this.point)

        return box
    }

    companion object {
        private val BOUNDING_KEY = toIndex("boundingbox")
    }
}
